### PR TITLE
Server Channel configuration and HTTP1 timeout

### PR DESCRIPTION
**Pull request by adam-fowler**
- Add configuration struct for HTTP1 and HTTP2 channels
- Add idle timeout parameter for HTTP1
- Add HTTP1 config to HTTP2 config for stream channels

I intend to use the HTTP2 config to add main channel support, like idle, ping timeout, close timeout etc

### CODE REVIEW

**Review by Joannis — APPROVED** [review]
Good idea, this will allow us to add further configuration as well.
Speaking of which, we should add some detail to our additionalChannelHandlers. Do they get added _before_ or _after_ the HTTP handlers?

File: Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift [hunk 1]
```diff
@@ -22,15 +22,48 @@ import NIOHTTPTypesHTTP1
 public struct HTTP1Channel: ServerChildChannel, HTTPChannelHandler {
     public typealias Value = NIOAsyncChannel<HTTPRequestPart, HTTPResponsePart>
 
+    /// HTTP1Channel configuration
+    public struct Configuration: Sendable {
+        /// Additional channel handlers to add to channel
```
Comment (Joannis) [inline Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift:27]: Can we document where these handlers are added?

File: Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift [hunk 2]
```diff
@@ -22,15 +22,48 @@ import NIOHTTPTypesHTTP1
...
+        public var additionalChannelHandlers: @Sendable () -> [any RemovableChannelHandler]
+        /// Time before closing an idle channel
+        public var idleTimeout: TimeAmount?
+
+        ///  Initialize HTTP1Channel.Configuration
+        /// - Parameters:
+        ///   - additionalChannelHandlers: Additional channel handlers to add to channel
```
Comment (Joannis) [inline Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift:34]: Can we document where these handlers are added?

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift [hunk 3]
```diff
@@ -30,26 +30,57 @@ public struct HTTP2UpgradeChannel: HTTPChannelHandler {
...
     }
 
+    /// HTTP2 Upgrade configuration
+    public struct Configuration: Sendable {
+        /// Configuration applieds to HTTP2 stream channels
+        public var streamConfiguration: HTTP1Channel.Configuration
+
+        ///  Initialize HTTP2UpgradeChannel.Configuration
+        /// - Parameters:
+        ///   - additionalChannelHandlers: Additional channel handlers to add to HTTP2 connection channel
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2Channel.swift:40]: Can we document where these handlers are added?

File: Sources/HummingbirdHTTP2/HTTPServerBuilder+http2.swift [hunk 4]
```diff
@@ -30,9 +30,10 @@ extension HTTPServerBuilder {
     ///   - tlsConfiguration: TLS configuration
     ///   - additionalChannelHandlers: Additional channel handlers to call before handling HTTP
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTPServerBuilder+http2.swift:31]: Can we document where these handlers are added?

File: Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift [hunk 5]
```diff
@@ -22,15 +22,48 @@ import NIOHTTPTypesHTTP1
 public struct HTTP1Channel: ServerChildChannel, HTTPChannelHandler {
     public typealias Value = NIOAsyncChannel<HTTPRequestPart, HTTPResponsePart>
 
+    /// HTTP1Channel configuration
+    public struct Configuration: Sendable {
+        /// Additional channel handlers to add to channel
+        public var additionalChannelHandlers: @Sendable () -> [any RemovableChannelHandler]
```
Comment (Joannis) [inline Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift:28]: Actually: Maybe this property can be clarified to reflect the position of the handlers?

File: Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift [hunk 6]
```diff
@@ -22,15 +22,48 @@ import NIOHTTPTypesHTTP1
...
+        /// Time before closing an idle channel
+        public var idleTimeout: TimeAmount?
+
+        ///  Initialize HTTP1Channel.Configuration
+        /// - Parameters:
+        ///   - additionalChannelHandlers: Additional channel handlers to add to channel
+        ///   - idleTimeout: Time before closing an idle channel
+        public init(
+            additionalChannelHandlers: @autoclosure @escaping @Sendable () -> [any RemovableChannelHandler] = [],
```
Comment (Joannis) [inline Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift:37]: Likewise, this property name can now maybe reflect the _position_ of the handlers?

**Review by Joannis — CHANGES_REQUESTED** [review]

**Comment by Joannis** [comment]
Let's discuss this change a bit more, I think we can improve it a bit more

File: Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift:27]: "Additional channel handlers to add to channel between HTTP part decode and HTTP request processing"?

Comment (Joannis) [reply Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift:27]: "Additional channel handlers to add to channel after HTTP part **decoding** and before HTTP request processing"

That sounds good

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift [hunk 7]
```diff
@@ -30,26 +30,57 @@ public struct HTTP2UpgradeChannel: HTTPChannelHandler {
...
+        ///   - streamConfiguration: Configuration applieds to HTTP2 stream channels
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2Channel.swift:41]:
```suggestion
        ///   - streamConfiguration: Configuration applied to HTTP2 stream channels
```

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift [hunk 8]
```diff
@@ -30,26 +30,57 @@ public struct HTTP2UpgradeChannel: HTTPChannelHandler {
         public let channel: Channel
     }
 
+    /// HTTP2 Upgrade configuration
+    public struct Configuration: Sendable {
+        /// Configuration applieds to HTTP2 stream channels
```
Comment (Joannis) [inline Sources/HummingbirdHTTP2/HTTP2Channel.swift:35]:
```suggestion
        /// Configuration applied to HTTP2 stream channels
```

File: Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift (hunk 1 above)
Comment (adam-fowler) [reply Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift]: Updated comment

File: Sources/HummingbirdHTTP2/HTTP2Channel.swift (hunk 3 above)
Comment (adam-fowler) [reply Sources/HummingbirdHTTP2/HTTP2Channel.swift:40]: This is fixed in the HTTP2 PR

File: Sources/HummingbirdHTTP2/HTTPServerBuilder+http2.swift (hunk 4 above)
Comment (adam-fowler) [reply Sources/HummingbirdHTTP2/HTTPServerBuilder+http2.swift:31]: This is fixed in the HTTP2 PR

File: Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift (hunk 5 above)
Comment (adam-fowler) [reply Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift:28]: Fixed comment

File: Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift (hunk 6 above)
Comment (adam-fowler) [reply Sources/HummingbirdCore/Server/HTTP/HTTP1Channel.swift:38]: Fixed comment

**Review by Joannis — APPROVED** [review]

**Pull request merged**